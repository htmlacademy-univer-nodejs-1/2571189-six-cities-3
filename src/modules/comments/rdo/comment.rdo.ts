import { Expose, Type } from 'class-transformer';
import UserRdo from '../../user/rdo/user.rdo.js';
import {UserType} from '../../../types/user.type.js';
export default class CommentRdo {
  @Expose()
  public id!: string;

  @Expose()
  public text!: string;

  @Expose()
  public rating!: number;

  @Expose({ name: 'createdAt'})
  public postDate!: string;

  @Expose({name: 'userId'})
  @Type(() => UserRdo)
    user!: UserType;
}
