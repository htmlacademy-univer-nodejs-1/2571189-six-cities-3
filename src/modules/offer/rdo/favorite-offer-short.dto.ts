import { Expose } from 'class-transformer';
import {CityEnum} from '../../../types/city.enum.js';
import {HousingType} from '../../../types/housing-type.enum.js';

export class FavoriteOfferShortDto {
  @Expose()
  public id!: string;

  @Expose()
    name!: string;

  @Expose({ name: 'createdAt'})
    publicationDate!: Date;

  @Expose()
    description!: string;

  @Expose()
    city!: CityEnum;

  @Expose()
    previewImage!: string;

  @Expose()
    premium!: boolean;

  favorite = true;

  @Expose()
    rating!: number;

  @Expose()
    housingType!: HousingType;

  @Expose()
    cost!: number;

  @Expose()
    commentsCount!: number;
}
