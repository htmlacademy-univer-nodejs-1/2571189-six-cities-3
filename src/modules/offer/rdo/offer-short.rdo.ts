import { Expose } from 'class-transformer';
import {CityEnum} from '../../../types/city.enum.js';
import {HousingType} from '../../../types/housing-type.enum.js';

export class OfferShortRdo {
  @Expose()
  public id!: string;

  @Expose()
    name!: string;

  @Expose({ name: 'createdAt'})
    publicationDate!: Date;

  @Expose()
    city!: CityEnum;

  @Expose()
    previewImage!: string;

  @Expose()
    premium!: boolean;

  @Expose()
    favorite!: boolean;

  @Expose()
    rating!: number;

  @Expose()
    housingType!: HousingType;

  @Expose()
    cost!: number;

  @Expose()
    commentsCount!: number;
}
