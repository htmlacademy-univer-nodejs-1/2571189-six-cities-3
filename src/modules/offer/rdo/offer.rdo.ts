import {Expose, Type} from 'class-transformer';
import {CityEnum} from '../../../types/city.enum.js';
import {HousingType} from '../../../types/housing-type.enum.js';
import {Facilities} from '../../../types/facilities.enum.js';
import {UserType} from '../../../types/user.type.js';
import {CoordinatesType} from '../../../types/coordinates.type.js';
import UserRdo from '../../user/rdo/user.rdo.js';

export class OfferRdo {
  @Expose()
  public id!: string;

  @Expose()
    name!: string;

  @Expose({ name: 'createdAt'})
    publicationDate!: Date;

  @Expose()
    description!: string;

  @Expose()
    city!: CityEnum;

  @Expose()
    previewImage!: string;

  @Expose()
    images!: string[];

  @Expose()
    premium!: boolean;

  @Expose()
    favorite!: boolean;

  @Expose()
    rating!: number;

  @Expose()
    housingType!: HousingType;

  @Expose()
    cost!: number;

  @Expose()
    commentsCount!: number;

  @Expose()
    roomCount!: number;

  @Expose()
    guestCount!: number;

  @Expose()
    facilities!: Facilities[];

  @Expose({name: 'userId'})
  @Type(() => UserRdo)
    offerAuthor!: UserType;

  @Expose()
    coordinates!: CoordinatesType;
}
