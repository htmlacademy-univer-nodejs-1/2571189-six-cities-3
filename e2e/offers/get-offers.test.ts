import assert from 'node:assert';
import { Headers, fetch } from 'undici';
import { describe, test } from 'vitest';
import {CityEnum} from '../../src/types/city.enum.js';
import {HousingType} from '../../src/types/housing-type.enum.js';
import {Facilities} from '../../src/types/facilities.enum.js';
import LoggedUserRdo from '../../src/modules/user/rdo/logged-user.rdo.js';
import UserRdo from '../../src/modules/user/rdo/user.rdo.js';
import {OfferRdo} from '../../src/modules/offer/rdo/offer.rdo.js';

process.env['E2E_ENDPOINT'] = 'http://localhost:4000';
assert(process.env['E2E_ENDPOINT'] !== undefined);
assert(process.env['E2E_ENDPOINT'].startsWith('http') === true, 'E2E_ENDPOINT должен начинаться с протокола');

const url = new URL(process.env['E2E_ENDPOINT']);
const OFFERS_COUNT = 3;

describe('GET /offers', async () => {
  test('Success get offers empty', async (tc) => {
    const user = {
      email: 'test@test.com',
      password: '12345678910',
    };

    await fetch(new URL('/users/register', url), {
      method: 'POST',
      headers: new Headers({
        'content-type': 'application/json',
      }),
      body: JSON.stringify({
        type: 'simple',
        email: 'test@test.com',
        username: 'name',
        password: '12345678910',
      })
    });

    const loginResponse = await fetch(new URL('/users/login', url), {
      method: 'POST',
      headers: new Headers({
        'content-type': 'application/json',
      }),
      body: JSON.stringify(user)
    });

    const content: LoggedUserRdo = await loginResponse.json() as unknown as LoggedUserRdo;
    const response = await fetch(new URL('/offers', url), {
      method: 'GET',
      headers: new Headers({
        'content-type': 'application/json',
        'Authorization': `Bearer ${content.token}`
      })
    });

    tc.expect(response.ok).toBeTruthy();
    tc.expect(response.status).toStrictEqual(200);
    tc.expect(response.headers.get('content-type')).toMatch(/application\/json/);
    const responseJson = await response.json() as OfferRdo[];
    tc.expect(responseJson.length).toStrictEqual(0);
  });
  test('Success get offers', async (tc) => {
    const user = {
      email: 'test1@test.com',
      password: '12345678910',
    };

    const registerResponse = await fetch(new URL('/users/register', url), {
      method: 'POST',
      headers: new Headers({
        'content-type': 'application/json',
      }),
      body: JSON.stringify({
        type: 'simple',
        email: 'test1@test.com',
        username: 'name1',
        password: '12345678910',
      })
    });

    const userRdo = await registerResponse.json() as unknown as UserRdo;
    const loginResponse = await fetch(new URL('/users/login', url), {
      method: 'POST',
      headers: new Headers({
        'content-type': 'application/json',
      }),
      body: JSON.stringify(user)
    });

    const content: LoggedUserRdo = await loginResponse.json() as unknown as LoggedUserRdo;
    for (let i = 0; i < OFFERS_COUNT; i++) {
      const offer = {
        name: `nametenlenght${i}`,
        description: 'descriptiontwentylenght',
        city: CityEnum.Amsterdam,
        premium: true,
        housingType: HousingType.House,
        roomCount: 2,
        guestCount: 3,
        cost: 10000,
        facilities: [Facilities.AirConditioning],
        userId: '',
        coordinates: {latitude: 2.22222, longitude: 3.444444}
      };

      const createResponse = await fetch(new URL('/offers', url), {
        method: 'POST',
        headers: new Headers({
          'content-type': 'application/json',
          'Authorization': `Bearer ${content.token}`
        }),
        body: JSON.stringify({...offer, userId: userRdo.id})
      });

      tc.expect(createResponse.ok).toBeTruthy();
    }

    const response = await fetch(new URL('/offers', url), {
      method: 'GET',
      headers: new Headers({
        'content-type': 'application/json',
        'Authorization': `Bearer ${content.token}`
      })
    });

    tc.expect(response.ok).toBeTruthy();
    tc.expect(response.status).toStrictEqual(200);
    tc.expect(response.headers.get('content-type')).toMatch(/application\/json/);
    const responseJson = await response.json() as OfferRdo[];
    const res = responseJson.map((x) => {
      return {...x, id: 0};
    });
    tc.expect(res).toMatchSnapshot();
    tc.expect(responseJson.length).toStrictEqual(OFFERS_COUNT);
  });

  test('Fail get offers [Unauthorized]', async (tc) => {
    const response = await fetch(new URL('/offers', url), {
      method: 'GET',
      headers: new Headers({
        'content-type': 'application/json',
        'Authorization': `Bearer ${123}`
      })
    });
    tc.expect(response.ok).toBeFalsy();
    tc.expect(response.status).toStrictEqual(401);
  });
});
