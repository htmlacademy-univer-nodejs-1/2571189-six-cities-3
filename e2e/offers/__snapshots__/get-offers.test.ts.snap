// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GET /offers > Success get offers 1`] = `
[
  {
    "city": "Amsterdam",
    "commentsCount": 0,
    "cost": 10000,
    "housingType": "House",
    "id": 0,
    "name": "nametenlenght2",
    "premium": true,
    "previewImage": "http://localhost:4000/upload/",
    "publicationDate": "2023-12-10T13:22:27.718Z",
    "rating": 1,
  },
  {
    "city": "Amsterdam",
    "commentsCount": 0,
    "cost": 10000,
    "housingType": "House",
    "id": 0,
    "name": "nametenlenght1",
    "premium": true,
    "previewImage": "http://localhost:4000/upload/",
    "publicationDate": "2023-12-10T13:22:27.697Z",
    "rating": 1,
  },
  {
    "city": "Amsterdam",
    "commentsCount": 0,
    "cost": 10000,
    "housingType": "House",
    "id": 0,
    "name": "nametenlenght0",
    "premium": true,
    "previewImage": "http://localhost:4000/upload/",
    "publicationDate": "2023-12-10T13:22:27.677Z",
    "rating": 1,
  },
]
`;
